File: week2/sync_vs_async.py
import asyncio
import os
import statistics
import time

import psutil

import config
import models
import utils

PROMPTS = config.BENCHMARK_PROMPTS


def _peak_memory_mb():
    return psutil.Process(os.getpid()).memory_info().rss / (1024 * 1024)


def run_sync():
    """Sequential baseline: one prompt at a time, each call blocks until it returns."""
    results = []
    start = time.perf_counter()
    for prompt in PROMPTS:
        chat = models.get_chat_model(config.BENCHMARK_MODEL, temperature=0.5)
        call_start = time.perf_counter()
        try:
            chat.invoke(prompt)
            results.append({"prompt": prompt, "elapsed": time.perf_counter() - call_start, "error": None})
        except Exception as error:
            results.append({"prompt": prompt, "elapsed": time.perf_counter() - call_start, "error": str(error)})
    total = time.perf_counter() - start
    return {"mode": "sync", "results": results, "total_seconds": total, "peak_memory_mb": _peak_memory_mb()}


async def run_async():
    """Concurrent version: all prompts in flight at once, capped by a semaphore so we
    don't blow through OpenRouter's free-tier per-minute request cap (see config.py)."""
    semaphore = asyncio.Semaphore(config.ASYNC_CONCURRENCY)

    async def call(prompt):
        async with semaphore:
            chat = models.get_chat_model(config.BENCHMARK_MODEL, temperature=0.5)
            call_start = time.perf_counter()
            try:
                await chat.ainvoke(prompt)
                return {"prompt": prompt, "elapsed": time.perf_counter() - call_start, "error": None}
            except Exception as error:
                return {"prompt": prompt, "elapsed": time.perf_counter() - call_start, "error": str(error)}

    start = time.perf_counter()
    results = await asyncio.gather(*(call(prompt) for prompt in PROMPTS))
    total = time.perf_counter() - start
    return {"mode": "async", "results": list(results), "total_seconds": total, "peak_memory_mb": _peak_memory_mb()}


def summarize(run):
    ok_times = [r["elapsed"] for r in run["results"] if not r["error"]]
    n_ok = len(ok_times)
    n_err = len(run["results"]) - n_ok
    avg_latency = statistics.mean(ok_times) if ok_times else 0.0
    throughput = n_ok / run["total_seconds"] if run["total_seconds"] else 0.0
    return {
        "mode": run["mode"],
        "total_seconds": round(run["total_seconds"], 2),
        "avg_latency_seconds": round(avg_latency, 2),
        "throughput_per_second": round(throughput, 3),
        "successes": n_ok,
        "errors": n_err,
        "peak_memory_mb": round(run["peak_memory_mb"], 1),
    }


def main():
    utils.console.print(
        f"\n[bold cyan]Exercise 3: Sync vs Async ({len(PROMPTS)} prompts, "
        f"model={config.BENCHMARK_MODEL}, async concurrency={config.ASYNC_CONCURRENCY})[/bold cyan]\n"
    )

    utils.console.print("Running SYNC (sequential) benchmark...")
    sync_summary = summarize(run_sync())

    utils.console.print("Running ASYNC (concurrent) benchmark...")
    async_summary = summarize(asyncio.run(run_async()))

    utils.print_benchmark_table(sync_summary, async_summary)

    if async_summary["total_seconds"] > 0:
        speedup = sync_summary["total_seconds"] / async_summary["total_seconds"]
    else:
        speedup = 0.0

    utils.console.print(
        f"\nAsync completed all {len(PROMPTS)} prompts [bold]{speedup:.2f}x[/bold] faster in wall-clock "
        f"time than sync (total_seconds: {sync_summary['total_seconds']}s sync vs "
        f"{async_summary['total_seconds']}s async), using at most {config.ASYNC_CONCURRENCY} requests "
        "in flight at once.\n"
        "Note: throughput here is dominated by network wait time (each call is I/O-bound - waiting "
        "on OpenRouter), not CPU. That's exactly why async helps: while one request is waiting on the "
        "network, the event loop can start the next one instead of sitting idle. Peak memory is included "
        "for completeness, but for this workload it stays roughly flat between the two modes - the win "
        "is entirely in wall-clock time, not resource usage.\n"
    )


if __name__ == "__main__":
    main()

File: week2-js/src/syncVsAsync.ts
/**
 * Exercise 3: sync (sequential) vs async (concurrent) throughput comparison.
 * Run: npm run benchmark
 */

import * as config from "./config.js";
import * as models from "./models.js";
import { printBenchmarkTable, type BenchmarkSummary } from "./utils.js";

const PROMPTS = config.BENCHMARK_PROMPTS;

interface CallResult {
  prompt: string;
  elapsed: number;
  error: string | null;
}

interface RunResult {
  mode: string;
  results: CallResult[];
  totalSeconds: number;
  peakMemoryMb: number;
}

function peakMemoryMb(): number {
  return process.memoryUsage().rss / (1024 * 1024);
}

/** Sequential baseline: one prompt at a time, each call blocks (via await) before the next starts. */
async function runSync(): Promise<RunResult> {
  const results: CallResult[] = [];
  const start = performance.now();
  for (const prompt of PROMPTS) {
    const chat = models.getChatModel(config.BENCHMARK_MODEL, { temperature: 0.5 });
    const callStart = performance.now();
    try {
      await chat.invoke(prompt);
      results.push({ prompt, elapsed: (performance.now() - callStart) / 1000, error: null });
    } catch (error) {
      results.push({ prompt, elapsed: (performance.now() - callStart) / 1000, error: String(error) });
    }
  }
  const total = (performance.now() - start) / 1000;
  return { mode: "sync", results, totalSeconds: total, peakMemoryMb: peakMemoryMb() };
}

/** Simple counting semaphore - the JS analog of asyncio.Semaphore, used below to cap
 * concurrency so we don't blow through OpenRouter's free-tier per-minute request cap. */
class Semaphore {
  private available: number;
  private readonly waiters: (() => void)[] = [];

  constructor(count: number) {
    this.available = count;
  }

  async acquire(): Promise<void> {
    if (this.available > 0) {
      this.available -= 1;
      return;
    }
    await new Promise<void>((resolve) => this.waiters.push(resolve));
  }

  release(): void {
    const next = this.waiters.shift();
    if (next) {
      next();
    } else {
      this.available += 1;
    }
  }
}

/** Concurrent version: all prompts in flight at once, capped by a semaphore. */
async function runAsync(): Promise<RunResult> {
  const semaphore = new Semaphore(config.ASYNC_CONCURRENCY);

  async function call(prompt: string): Promise<CallResult> {
    await semaphore.acquire();
    try {
      const chat = models.getChatModel(config.BENCHMARK_MODEL, { temperature: 0.5 });
      const callStart = performance.now();
      try {
        await chat.invoke(prompt);
        return { prompt, elapsed: (performance.now() - callStart) / 1000, error: null };
      } catch (error) {
        return { prompt, elapsed: (performance.now() - callStart) / 1000, error: String(error) };
      }
    } finally {
      semaphore.release();
    }
  }

  const start = performance.now();
  const results = await Promise.all(PROMPTS.map((prompt) => call(prompt)));
  const total = (performance.now() - start) / 1000;
  return { mode: "async", results, totalSeconds: total, peakMemoryMb: peakMemoryMb() };
}

function summarize(run: RunResult): BenchmarkSummary {
  const okTimes = run.results.filter((r) => !r.error).map((r) => r.elapsed);
  const nOk = okTimes.length;
  const nErr = run.results.length - nOk;
  const avgLatency = okTimes.length ? okTimes.reduce((a, b) => a + b, 0) / okTimes.length : 0;
  const throughput = run.totalSeconds ? nOk / run.totalSeconds : 0;
  return {
    mode: run.mode,
    total_seconds: Math.round(run.totalSeconds * 100) / 100,
    avg_latency_seconds: Math.round(avgLatency * 100) / 100,
    throughput_per_second: Math.round(throughput * 1000) / 1000,
    successes: nOk,
    errors: nErr,
    peak_memory_mb: Math.round(run.peakMemoryMb * 10) / 10,
  };
}

async function main(): Promise<void> {
  console.log(
    `\nExercise 3: Sync vs Async (${PROMPTS.length} prompts, model=${config.BENCHMARK_MODEL}, async concurrency=${config.ASYNC_CONCURRENCY})\n`
  );

  console.log("Running SYNC (sequential) benchmark...");
  const syncSummary = summarize(await runSync());

  console.log("Running ASYNC (concurrent) benchmark...");
  const asyncSummary = summarize(await runAsync());

  printBenchmarkTable(syncSummary, asyncSummary);

  const speedup = asyncSummary.total_seconds > 0 ? syncSummary.total_seconds / asyncSummary.total_seconds : 0;

  console.log(
    `\nAsync completed all ${PROMPTS.length} prompts ${speedup.toFixed(2)}x faster in wall-clock ` +
      `time than sync (total_seconds: ${syncSummary.total_seconds}s sync vs ` +
      `${asyncSummary.total_seconds}s async), using at most ${config.ASYNC_CONCURRENCY} requests ` +
      "in flight at once.\n" +
      "Note: throughput here is dominated by network wait time (each call is I/O-bound - waiting " +
      "on OpenRouter), not CPU. That's exactly why concurrency helps: while one request is waiting on " +
      "the network, Node's event loop can start the next one instead of sitting idle. Peak memory is " +
      "included for completeness, but for this workload it stays roughly flat between the two modes - " +
      "the win is entirely in wall-clock time, not resource usage.\n"
  );
}

main();
